neps/nep-0384.md

```diff
@@ -0,0 +1,108 @@
+---
+NEP: 384
+Title: Host function for fetching block hashes in contract runtime
+Author: Seun Lanlege <seunlanlege@gmail.com>
+DiscussionsTo: https://github.com/near/NEPs/pull/384
+Status: Draft
+Type: Runtime Spec
+Category: Contract
+Created: 27-Aug-2022
+---
+
+## Summary
+
+A new host function `env::block_hash` which will allow contract authors access the blockhashes of 256 of the most reccent blocks. This is akin to the [`blockhash`](https://docs.soliditylang.org/en/v0.8.15/units-and-global-variables.html#block-and-transaction-properties) api in ethereum.
```

Comment:
One epoch is 43_200 blocks and RPC nodes are configured to keep a 5 epochs worth of data before garbage collecting.  So I would imagine that a RPC node can easily return information about more than just 256 recent blocks.  Should this number be increased or specified in a different way to make the feature more generic?

Reply:
> So I would imagine that a RPC node can easily return information about more than just 256 recent blocks.

All nodes in the network must have access to those values, otherwise any block containing transactions that access much older block hashes would panic

Reply:
> One epoch is 43_200 blocks and RPC nodes are configured to keep a 5 epochs worth of data before garbage collecting.

I don’t think we can relay on this at protocol lever though.  For example, RPC nodes can be configured to store just three epochs worth of data.  Even that is partially because we settled on arbitrary limit of at least three epochs.  In practice RPC node could work with a shorter horizon I think.  I’d rather have this function be limited to as few blocks as possible so we don’t encode current client configuration in the protocol.

The question here is, where does the 256 comes from?

Reply:
> 256

This will have to be enforced at the protocol level, as minimum number of stored blocks to validate inncoming blocks